jicofo/src/main/kotlin/org/jitsi/jicofo/conference/Participant.kt

```diff
@@ -0,0 +1,375 @@
+    }
+
+    /**
+     * Schedule a task to signal all queued remote sources to the remote side. If a task is already scheduled, does
+     * not schedule a new one (the existing task will send all latest queued sources).
+     */
+    private fun scheduleSignalingOfQueuedSources() {
+        val delayMs = ConferenceConfig.config.getSourceSignalingDelayMs(conference.participantCount)
+        synchronized(signalQueuedSourcesTaskSyncRoot) {
+            if (signalQueuedSourcesTask == null) {
+                logger.debug("Scheduling a task to signal queued remote sources after $delayMs ms.")
+                signalQueuedSourcesTask = scheduledPool.schedule(
+                    {
+                        synchronized(signalQueuedSourcesTaskSyncRoot) {
+                            sendQueuedRemoteSources()
+                            signalQueuedSourcesTask = null
+                        }
+                    },
+                    delayMs.toLong(),
+                    TimeUnit.MILLISECONDS
+                )
+            }
+        }
+    }
+
+    /**
+     * Remove a set of remote sources, which are to be signaled as removed to the remote side. The sources may be
+     * signaled immediately, or queued to be signaled later.
+     */
+    fun removeRemoteSources(sources: ConferenceSourceMap?) {
+        if (useSsrcRewriting()) {
+            // Bridge will signal sources in this case.
+            return
+        }
+        synchronized(sourceSignaling) { sourceSignaling.removeSources(sources!!) }
+        if (jingleSession == null) {
+            logger.debug("No Jingle session yet, queueing source-remove.")
+            // No need to schedule, the sources will be signaled when the session is established.
+            return
+        }
+        synchronized(signalQueuedSourcesTaskSyncRoot) { scheduleSignalingOfQueuedSources() }
+    }
+
+    /**
+     * Signal any queued remote source modifications (either addition or removal) to the remote side.
+     */
+    fun sendQueuedRemoteSources() {
+        val jingleSession = jingleSession
+        if (jingleSession == null) {
+            logger.warn("Can not signal remote sources, Jingle session not established.")
+            return
+        }
+        for ((action, sources) in sourceSignaling.update()) {
+            logger.info(
+                "Sending a queued source-" + action.toString().lowercase(Locale.getDefault()) + ", sources:" + sources
+            )
+            if (action === AddOrRemove.Add) {
+                jingleSession.addSource(sources)
+            } else if (action === AddOrRemove.Remove) {
+                jingleSession.removeSource(sources)
+            }
+        }
+    }
+
+    /**
+     * Whether force-muting should be suppressed for this participant (it is a trusted participant and doesn't
+     * support unmuting).
+     */
+    fun shouldSuppressForceMute() = chatMember.isJigasi && !hasAudioMuteSupport() || chatMember.isJibri
```

Comment:
I'd put the parens back here, remembering the relative priority of `&&` and `||` is unnecessary cognitive load